File: R/tutorial/basic-examples/merge_artifacts_flow.R
library(metaflow)

start <- function(self){
    self$pass_down <- "non-modified"
}

a <- function(self){
    self$common <- "common in a and b"
    self$x <- "x in a" 
    self$y <- "y in a" 
    self$from_a <- "only in a" 
}

b <- function(self){
    self$common <- "common in a and b" 
    self$x <- "x in b" 
    self$y <- "y in b" 
}

join <- function(self, inputs){
    # manually propogate variable that has different values in different branches 
    self$x <- inputs[[1]]$x

    merge_artifacts(self, inputs, exclude=list("y"))

    # If without the merge_artifact, the following artifact access won't work. 
    print(paste('pass_down is', self$pass_down))
    print(paste('from_a is', self$from_a))
    print(paste('common is', self$common))
}

metaflow("BranchFlow") %>%
    step(step = "start",
         r_function=start,
         next_step = c("a", "b")) %>%
    step(step = "a", 
         r_function=a, 
         next_step="join") %>%
    step(step="b", 
         r_function=b, 
         next_step="join") %>%
    step(step="join", 
         r_function=join, 
         next_step="end",
         join=TRUE) %>%
    step(step="end") %>%
    run()

File: test/unit/spin/flows/merge_artifacts_flow.py
from metaflow import FlowSpec, step


class MergeArtifactsFlow(FlowSpec):

    @step
    def start(self):
        self.pass_down = "a"
        self.next(self.a, self.b)

    @step
    def a(self):
        self.common = 5
        self.x = 1
        self.y = 3
        self.from_a = 6
        self.next(self.join)

    @step
    def b(self):
        self.common = 5
        self.x = 2
        self.y = 4
        self.next(self.join)

    @step
    def join(self, inputs):
        self.x = inputs.a.x
        self.merge_artifacts(inputs, exclude=["y"])
        print("x is %s" % self.x)
        print("pass_down is %s" % self.pass_down)
        print("common is %d" % self.common)
        print("from_a is %d" % self.from_a)
        self.next(self.c)

    @step
    def c(self):
        self.next(self.d, self.e)

    @step
    def d(self):
        self.conflicting = 7
        self.next(self.join2)

    @step
    def e(self):
        self.conflicting = 8
        self.next(self.join2)

    @step
    def join2(self, inputs):
        self.merge_artifacts(inputs, include=["pass_down", "common"])
        print("Only pass_down and common exist here")
        self.next(self.end)

    @step
    def end(self):
        pass


if __name__ == "__main__":
    MergeArtifactsFlow()
